### PR TITLE
[SCB-2333] HttpServerOptions.maxFormAttributeSize can be configurable

### PR DESCRIPTION
**Pull request by david6969xin**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean install -Pit` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

File: transports/transport-rest/transport-rest-vertx/src/main/java/org/apache/servicecomb/transport/rest/vertx/RestServerVerticle.java [hunk 1]
```diff
@@ -253,10 +253,22 @@ private HttpServerOptions createDefaultHttpServerOptions() {
     serverOptions.setIdleTimeout(TransportConfig.getConnectionIdleTimeoutInSeconds());
     serverOptions.setCompressionSupported(TransportConfig.getCompressed());
     serverOptions.setMaxHeaderSize(TransportConfig.getMaxHeaderSize());
+    serverOptions.setMaxFormAttributeSize(TransportConfig.getFormAttributeSize());
```
Comment (liubao68) [inline transports/transport-rest/transport-rest-vertx/src/main/java/org/apache/servicecomb/transport/rest/vertx/RestServerVerticle.java:256]: maybe `getMaxFormAttributeSize` method name is better.

File: transports/transport-rest/transport-rest-vertx/src/main/java/org/apache/servicecomb/transport/rest/vertx/RestServerVerticle.java [hunk 2]
```diff
@@ -253,10 +253,22 @@ private HttpServerOptions createDefaultHttpServerOptions() {
...
+    serverOptions.setCompressionLevel(TransportConfig.getCompressionLevel());
+    serverOptions.setMaxChunkSize(TransportConfig.getMaxChunkSize());
+    serverOptions.setDecompressionSupported(TransportConfig.getDecompression());
```
Comment (liubao68) [inline transports/transport-rest/transport-rest-vertx/src/main/java/org/apache/servicecomb/transport/rest/vertx/RestServerVerticle.java:259]: maybe `getDecompressionSupported` method name is better.

Comment (david6969xin) [reply transports/transport-rest/transport-rest-vertx/src/main/java/org/apache/servicecomb/transport/rest/vertx/RestServerVerticle.java:259]: ok

File: transports/transport-rest/transport-rest-vertx/src/main/java/org/apache/servicecomb/transport/rest/vertx/RestServerVerticle.java (hunk 1 above)
Comment (david6969xin) [reply transports/transport-rest/transport-rest-vertx/src/main/java/org/apache/servicecomb/transport/rest/vertx/RestServerVerticle.java:256]: ok

File: demo/demo-springmvc/springmvc-server/src/main/java/org/apache/servicecomb/demo/springmvc/server/FormRequestSchema.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+@RequestMapping(path = "/form", produces = MediaType.APPLICATION_JSON_VALUE)
+public class FormRequestSchema {
+
+  @PostMapping(path = "/formRequest")
+  public String formRequest(@FormParam("formData") String formData, @FormParam("flag") String flag) throws Exception {
+    if (formData.getBytes().length <= (flag == null ? HttpServerOptions.DEFAULT_MAX_FORM_ATTRIBUTE_SIZE
+        : TransportConfig.getMaxFormAttributeSize())) {
+      return "formRequest success";
+    }
+    return null;
```
Comment (liubao68) [inline demo/demo-springmvc/springmvc-server/src/main/java/org/apache/servicecomb/demo/springmvc/server/FormRequestSchema.java:42]: return "formRequest success";

File: demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/TestFormRequestSchema.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  // formSize is greater than default maxFormAttributeSize , throw exception
+  private void testFormRequestFail() throws Exception {
+    RestTemplate restTemplate = RestTemplateBuilder.create();
+    HttpHeaders headers = new HttpHeaders();
+    headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
+    MultiValueMap<String, String> formData = new LinkedMultiValueMap<>();
+    StringBuffer stringBuffer = new StringBuffer();
+    for (int i = 0; i < 3072; i++) {
```
Comment (liubao68) [inline demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/TestFormRequestSchema.java:68]: 4096

File: demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/TestFormRequestSchema.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      ResponseEntity<String> responseEntity = restTemplate
+          .postForEntity("servicecomb://springmvc/form/formRequest", requestEntity, String.class);
+      TestMgr.check(responseEntity.getBody(), null);
+    } catch (InvocationException e) {
+      TestMgr.check(e.getMessage().equals("Size exceed allowed maximum capacity"), true);
+    }
+  }
+
+  // formSize is greater than default maxFormAttributeSize ,success through external configuration
+  private void testFormRequestSuccess() throws Exception {
```
Comment (liubao68) [inline demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/TestFormRequestSchema.java:83]: This test case is useless

File: demo/demo-springmvc/springmvc-server/src/main/java/org/apache/servicecomb/demo/springmvc/server/FormRequestSchema.java (hunk 3 above)
Comment (david6969xin) [reply demo/demo-springmvc/springmvc-server/src/main/java/org/apache/servicecomb/demo/springmvc/server/FormRequestSchema.java]: done

File: demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/TestFormRequestSchema.java (hunk 4 above)
Comment (david6969xin) [reply demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/TestFormRequestSchema.java]: done

File: demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/TestFormRequestSchema.java (hunk 5 above)
Comment (david6969xin) [reply demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/TestFormRequestSchema.java]: ok，deleted

File: demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/TestFormRequestSchema.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
+    MultiValueMap<String, String> formData = new LinkedMultiValueMap<>();
+    StringBuffer stringBuffer = new StringBuffer();
+    for (int i = 0; i < 4096; i++) {
+      stringBuffer.append("a");
+    }
+    formData.add("formData", String.valueOf(stringBuffer));
+    HttpEntity<MultiValueMap<String, String>> requestEntity = new HttpEntity<>(formData, headers);
+    try {
+      restTemplate.postForEntity("servicecomb://springmvc/form/formRequest", requestEntity, String.class);
```
Comment (liubao68) [inline demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/TestFormRequestSchema.java:73]: fail if exception not thrown

File: demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/TestFormRequestSchema.java (hunk 4 above)
Comment (david6969xin) [reply demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/TestFormRequestSchema.java]: this value should be greater than configurable value，it just contain value size not key size

File: demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/TestFormRequestSchema.java (hunk 6 above)
Comment (david6969xin) [reply demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/TestFormRequestSchema.java:73]: done

**Review by liubao68 — APPROVED** [review]

**Pull request merged**